### PR TITLE
Raw image and supporting images

### PR DESCRIPTION
**Pull request by profxj**
After stewing on this for 24hrs, I decided on another approach which bundles all of the previous raw image methods into one `get_rawimage()` in the `Spectrograph` class.  It is less than 100 lines.  This passes back the 5 items important for image processing.  I nearly bundled those into a very lightweight object named `RawImage` but decided otherwise (for now).

This has the following pros and cons:

- Pro: One-stop shopping, i.e. easier to maintain than the multiple methods
- Pro: Reduces multiple loads of the image 
- Pro: When developing a new instrument, it is clearer what needs to be provided
- Con: If you want only one of the 5 items you can't get only it.  (e.g. BPM generation)
- Con: ??

Here are the specific CHANGES:

- Refactor BPM generation
- Merge raw_image loading with datasec_img and oscansec_img generation
- Sync datasec_img to image in ProcessRawImage
- Started (barely) on a path to having calibration images in counts and not ADU

Passing unit tests.  Failing GMOS in the dev suite but that will be 
fixed in a PR to follow right after.

### CODE REVIEW

**Comment by profxj** [comment]
(base) profx> py.test --disable-warnings 
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-4.3.1, py-1.8.0, pluggy-0.9.0
rootdir: /data/Projects/Python/PypeIt, inifile:
plugins: remotedata-0.3.1, openfiles-0.3.2, doctestplus-0.3.0, cov-2.7.1, arraydiff-0.3
collected 155 items                                                            

pypeit/tests/test_arc.py .                                               [  0%]
pypeit/tests/test_biasframe.py ....                                      [  3%]
pypeit/tests/test_bpmimage.py ..                                         [  4%]
pypeit/tests/test_calibrationimage.py .                                  [  5%]
pypeit/tests/test_calibrations.py ...........                            [ 12%]
pypeit/tests/test_coadd.py .                                             [ 12%]
pypeit/tests/test_cooked.py .                                            [ 13%]
pypeit/tests/test_flatfield.py .                                         [ 14%]
pypeit/tests/test_flexure.py .                                           [ 14%]
pypeit/tests/test_flux.py ....                                           [ 17%]
pypeit/tests/test_fluxspec.py ...                                        [ 19%]
pypeit/tests/test_load.py ..                                             [ 20%]
pypeit/tests/test_load_images.py ..........                              [ 27%]
pypeit/tests/test_masterframe.py .                                       [ 27%]
pypeit/tests/test_match.py ....                                          [ 30%]
pypeit/tests/test_metadata.py .....                                      [ 33%]
pypeit/tests/test_msgs.py ..                                             [ 34%]
pypeit/tests/test_opticalmodel.py ....                                   [ 37%]
pypeit/tests/test_parse.py ...                                           [ 39%]
pypeit/tests/test_processimage.py ..                                     [ 40%]
pypeit/tests/test_procimg.py .                                           [ 41%]
pypeit/tests/test_pydl.py .                                              [ 41%]
pypeit/tests/test_pypeit.py ..                                           [ 43%]
pypeit/tests/test_pypeitimage.py ..                                      [ 44%]
pypeit/tests/test_pypeitpar.py ......................                    [ 58%]
pypeit/tests/test_pypeitsetup.py .......                                 [ 63%]
pypeit/tests/test_qa.py .                                                [ 63%]
pypeit/tests/test_save.py ..                                             [ 65%]
pypeit/tests/test_scienceimage.py ....                                   [ 67%]
pypeit/tests/test_scripts.py ..                                          [ 69%]
pypeit/tests/test_setups.py .............                                [ 77%]
pypeit/tests/test_slitmask.py .                                          [ 78%]
pypeit/tests/test_specobj.py ..                                          [ 79%]
pypeit/tests/test_specobjs.py ....                                       [ 81%]
pypeit/tests/test_spectrographs.py ............                          [ 89%]
pypeit/tests/test_traceimage.py ..                                       [ 90%]
pypeit/tests/test_traceslits.py .....                                    [ 94%]
pypeit/tests/test_utils.py ..                                            [ 95%]
pypeit/tests/test_vcorr.py ..                                            [ 96%]
pypeit/tests/test_wavecalib.py .                                         [ 97%]
pypeit/tests/test_waveimage.py .                                         [ 98%]
pypeit/tests/test_wavemodel.py .                                         [ 98%]
pypeit/tests/test_wavetilts.py ..                                        [100%]

================== 155 passed, 109 warnings in 597.09 seconds ==================

**Comment by profxj** [comment]
Dev suite:

--- PYPEIT DEVELOPMENT SUITE FAILED 2/30 TESTS ---

One fail is GMOS (will be fixed in a PR I'll submit right after)
The other is due to a file missing in the Dev Suite Repo

File: pypeit/images/processrawimage.py [hunk 1]
```diff
@@ -98,6 +100,7 @@ def bpm(self):
                                     det=self.det)
         return self._bpm
 
+    '''
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/images/processrawimage.py]: removed that code

File: pypeit/core/procimg.py [hunk 2]
```diff
@@ -249,7 +249,7 @@ def gain_frame(amp_img, gain, trim=True):
             List of amplifier gain values.  Must be that the gain for
             amplifier 1 is provided by `gain[0]`, etc.
         trim (:obj:`bool`, optional):
-            Trim the overscan section from the image.
+            Trim; not recommended
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/procimg.py]: removed

File: pypeit/images/processrawimage.py [hunk 3]
```diff
@@ -342,9 +343,12 @@ def trim(self, force=False):
             msgs.warn("Image was already trimmed.  Returning current image")
             return self.image
         # Do it
-        trim_image = procimg.trim_frame(self.image, self.rawdatasec_img < 1)
+        trim_image = procimg.trim_frame(self.image, self.datasec_img < 1)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/images/processrawimage.py]: agreed. done.

File: pypeit/tests/test_bpmimage.py [hunk 4]
```diff
@@ -38,12 +39,15 @@ def test_keck_lris_red():
...
     det = 2
+    '''
     rdsec_img = spectrograph.get_rawdatasec_img(example_file, det=det)
     trim = procimg.trim_frame(rdsec_img, rdsec_img < 1)
     orient = spectrograph.orient_image(trim, det)
     shape = orient.shape
+    '''
     # Simple
-    bpm = spectrograph.bpm(shape=shape, filename=example_file, det=det)
+    #bpm = spectrograph.bpm(shape=shape, filename=example_file, det=det)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/tests/test_bpmimage.py]: agreed; removed (along with other commented code in that file)

File: pypeit/tests/test_load_images.py [hunk 5]
```diff
@@ -23,8 +23,11 @@ def ProcessImages(specstr, par, files, det=1):
     return calibImage
 
 def grab_img(proc, filename):
-    rdimg = proc.spectrograph.get_rawdatasec_img(filename, proc.det)
-    data_img, slice = procimg.rect_slice_with_mask(proc.image, rdimg)
+    #img, hdu = proc.spectrograph.load_raw_frame(filename, proc.det)
+    #exptime, rawdatasec_img, oscansec_img, binning_raw = proc.spectrograph.load_raw_extras(hdu, proc.det)
+    img, hdu, exptime, rawdatasec_img, oscansec_img = proc.spectrograph.get_rawimage(filename, proc.det)
+    #rdimg = proc.spectrograph.get_rawdatasec_img(filename, proc.det)
+    data_img, slice = procimg.rect_slice_with_mask(img, rawdatasec_img)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/tests/test_load_images.py]: yup, good catch.  changed.

File: pypeit/spectrographs/spectrograph.py [hunk 6]
```diff
@@ -110,19 +105,28 @@ def __init__(self):
...
+                If datasec_img is provided, an image with the same shape
+                is returned with the gain accounted for
+        """
+        nonlinear_counts = self.detector[det-1]['saturation']*self.detector[det-1]['nonlinear']
+        # Generate an image, applying the gain?
+        if datasec_img is not None:
+            gain = np.atleast_1d(self.detector[det-1]['gain']).tolist()
+            nonlinear_counts = nonlinear_counts * procimg.gain_frame(datasec_img, gain)
+        # Return
+        return nonlinear_counts
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/spectrographs/spectrograph.py]: This was a (minimal) start towards allowing for calibration
frames to be in counts and still deal with saturation.   So
this option isn't implemented yet.   And I'm going to add that
comment to the doc string.

File: pypeit/images/scienceimage.py [hunk 7]
```diff
@@ -45,6 +45,8 @@ class ScienceImage(pypeitimage.PypeItImage, maskimage.ImageMask):
             Inverse variance image
         rn2img (np.narray):
             Read noise**2 image
+        datasec_img (np.ndarray):
```
Comment (jhennawi) [inline pypeit/images/scienceimage.py]: I'm not following why the scienceimage needs to know about the datasec_img. The sciencimage should only operate on a processed gain multiplied image and this should take place in the call to process.

File: pypeit/images/scienceimage.py [hunk 8]
```diff
@@ -330,10 +337,11 @@ def build_ivar(self):
         # Convenience
         detector = self.spectrograph.detector[self.det-1]
         # For RN variations by Amp
-        datasec_img = self.spectrograph.get_datasec_img(filename=self.filename,
-                                                     det=self.det)
+        #embed(header='333 of sciencimage')
+        #datasec_img = self.spectrograph.get_datasec_img(filename=self.filename,
+        #                                             det=self.det)
         # Generate
```
Comment (jhennawi) [inline pypeit/images/scienceimage.py]: See my comment above. All these operations should be in one image processing routine. These steps have nothing to do with the image being a science image.

File: pypeit/images/processrawimage.py [hunk 9]
```diff
@@ -38,8 +38,11 @@ class ProcessRawImage(pypeitimage.PypeItImage):
             Dict describing the steps performed on the image
         _bpm (np.ndarray):
             Holds the bad pixel mask once loaded
```
Comment (jhennawi) [inline pypeit/images/processrawimage.py]: See my comment below. The noise model and read noise image should be generated in this function call, not scienceimage.

File: pypeit/spectrographs/keck_deimos.py [hunk 10]
```diff
@@ -394,7 +397,86 @@ def idname(self, ftype):
                  'standard': None,
                  'trace': 'IntFlat' }
         return name[ftype]
-  
+
```
Comment (jhennawi) [inline pypeit/spectrographs/keck_deimos.py]: Thanks for restructuring this. This is what I thin we want, a set of instrument specific image processing routines that overload generic routines in the generic spectrograph.py file with the same name.

File: pypeit/scripts/view_fits.py [hunk 11]
```diff
@@ -51,12 +51,14 @@ def main(args):
 
     # RAW_LRIS??
```
Comment (jhennawi) [inline pypeit/scripts/view_fits.py]: Can we structure this routine more generally to just require the spectrograph and then read in the image with the spectrograph class. I don't see why you need to specify raw_lris or raw_deimos anymore given how you restructured things. 

Even better would be if you figured out the instrument from the header thus not requiring the instrument argument.

File: pypeit/spectrographs/spectrograph.py [hunk 12]
```diff
@@ -110,19 +105,32 @@ def __init__(self):
     def default_pypeit_par():
         return pypeitpar.PypeItPar()
 
-    def nonlinear_counts(self, det=1):
+    def nonlinear_counts(self, det, datasec_img=None):
```
Comment (jhennawi) [inline pypeit/spectrographs/spectrograph.py]: Thanks for implementing this. See my suggestion above for what the processrawimage object should include. I think non-linear counts should be an attribute for the object that holds the images.

File: pypeit/images/scienceimage.py (hunk 7 above)
Comment (profxj) [reply pypeit/images/scienceimage.py]: This is a re-hash of Issues #740 and #719.  
As I mentioned in the closed PR, I started walking on this
path and found it too arduous for this PR.  Am not 
taking it on yet.

File: pypeit/images/scienceimage.py (hunk 8 above)
Comment (profxj) [reply pypeit/images/scienceimage.py]: Future PR

File: pypeit/images/processrawimage.py (hunk 9 above)
Comment (profxj) [reply pypeit/images/processrawimage.py]: Not this PR

File: pypeit/scripts/view_fits.py (hunk 11 above)
Comment (profxj) [reply pypeit/scripts/view_fits.py]: spectrograph now required

**Comment by profxj** [comment]
merging..

**Pull request merged**